gpytorch/likelihoods/likelihood.py

```diff
@@ -1,60 +1,134 @@
-        (co)variance of p(f|...).
+        Args:
+            :attr:`function_samples`
+                Samples from the function `f`
+            :attr:`kwargs`
+
+        Returns:
+            Distribution object (with same shape as :attr:`function_samples`)
         """
         raise NotImplementedError
 
-    def variational_log_probability(self, f, y):
+    def expected_log_prob(self, observations, function_dist, *params, **kwargs):
         """
-        Compute the log likelihood p(y|f) given y and averaged over a set of
-        latent function samples.
+        Computes the expected log likelihood (used for variational inference):
 
-        For the purposes of our variational inference implementation, y is an
-        n-by-1 label vector, and f is an n-by-s matrix of s samples from the
-        variational posterior, q(f|D).
+        .. math::
+            \mathbb{E}_{f(x)} \left[ \log p \left( y \mid f(x) \right) \right]
+
+        Args:
+            :attr:`function_dist` (:class:`gpytorch.distributions.MultivariateNormal`)
+                Distribution for :math:`f(x)`.
+            :attr:`observations` (:class:`torch.Tensor`)
+                Values of :math:`y`.
+            :attr:`kwargs`
+
+        Returns
+            `torch.Tensor` (log probability)
         """
-        raise NotImplementedError
+        log_prob_lambda = lambda function_samples: self.forward(function_samples).log_prob(observations)
+        log_prob = self.quadrature(log_prob_lambda, function_dist)
+        return log_prob.sum(tuple(range(len(function_dist.event_shape))))
 
-    def pyro_sample_y(self, variational_dist_f, y_obs, sample_shape, name_prefix=""):
-        raise NotImplementedError
+    def marginal(self, function_dist, *params, **kwargs):
+        """
+        Computes a predictive distribution :math:`p(y*|x*)` given either a posterior
+        distribution :math:`p(f|D,x)` or a prior distribution :math:`p(f|x)` as input.
+
+        With both exact inference and variational inference, the form of
+        :math:`p(f|D,x)` or :math:`p(f|x)` should usually be Gaussian. As a result, input
+        should usually be a MultivariateNormal specified by the mean and
+        (co)variance of :math:`p(f|...)`.
+
+        Args:
+            :attr:`function_dist` (:class:`gpytorch.distributions.MultivariateNormal`)
+                Distribution for :math:`f(x)`.
+            :attr:`kwargs`
+
+        Returns
+            Distribution object (the marginal distribution, or samples from it)
+        """
+        sample_shape = torch.Size((settings.num_likelihood_samples,))
+        function_samples = function_dist.rsample(sample_shape)
+        return self.forward(function_samples)
+
+    def __call__(self, input, *params, **kwargs):
+        # Conditional
+        if torch.is_tensor(input):
+            raise RuntimeError("Likelihoods expects a MultivariateNormal input to make predictions, got a {}".format(
```

Comment:
On reflection about my comment below, I'm assuming this runtime error just shouldn't be here since it's immediately followed by a call to forward?

Reply:
Yes - that was not intentional :)